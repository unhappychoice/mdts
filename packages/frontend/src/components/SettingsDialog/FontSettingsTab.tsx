import {
  Box,
  FormControlLabel,
  MenuItem,
  Radio,
  RadioGroup,
  Select,
  Slider,
  TextField,
  Typography
} from '@mui/material';
import React, { useCallback } from 'react';
import { WEB_SAFE_FONTS, WEB_SAFE_MONOSPACE_FONTS, } from '../../constants';

interface FontSettingsTabProps {
  fontSize: number;
  fontFamily: string;
  fontFamilyMonospace: string;
  fontInputMode: 'select' | 'text';
  fontInputModeMonospace: 'select' | 'text';
  setFontSize: (size: number) => void;
  setFontFamily: (font: string) => void;
  setFontFamilyMonospace: (font: string) => void;
  handleFontInputModeChange: (_: React.MouseEvent<HTMLElement>, newMode: 'select' | 'text' | null) => void;
  handleFontInputModeMonospaceChange: (_: React.MouseEvent<HTMLElement>, newMode: 'select' | 'text' | null) => void;
}

const FontSettingsTab: React.FC<FontSettingsTabProps> = ({
  fontSize,
  fontFamily,
  fontFamilyMonospace,
  fontInputMode,
  fontInputModeMonospace,
  setFontSize,
  setFontFamily,
  setFontFamilyMonospace,
  handleFontInputModeChange,
  handleFontInputModeMonospaceChange,
}) => {

  const handleFontSelect = useCallback((e) => {
    setFontFamily(e.target.value as string);
  }, [setFontFamily]);

  const handleFontChange = useCallback((e) => {
    setFontFamily(e.target.value);
  }, [setFontFamily]);

  const handleMonospaceFontSelect = useCallback((e) => {
    setFontFamilyMonospace(e.target.value as string);
  }, [setFontFamilyMonospace]);

  const handleMonospaceFontChange = useCallback((e) => {
    setFontFamilyMonospace(e.target.value);
  }, [setFontFamilyMonospace]);

  const handleChangeFontSize = useCallback((_, newValue) => {
    setFontSize(newValue as number);
  }, [setFontSize]);

  return (
    <Box sx={{ p: 2 }}>
      <Typography variant="subtitle2" gutterBottom>Font Size</Typography>
      <Slider
        value={fontSize}
        onChange={handleChangeFontSize}
        aria-label="Font Size"
        aria-labelledby="font-size-slider"
        valueLabelDisplay="auto"
        step={1}
        marks
        min={10}
        max={24}
        size="small"
        sx={{ mb: 1 }}
      />

      <Typography variant="subtitle2" gutterBottom sx={{ mt: 2 }} id="font-family-label">Font Family</Typography>
      <RadioGroup
        row
        aria-labelledby="font-family-label"
        name="font-input-mode-group"
        value={fontInputMode}
        onChange={handleFontInputModeChange}
        sx={{ mb: 1 }}
      >
        <FormControlLabel value="select" control={<Radio size="small" />} label="Select from list" />
        <FormControlLabel value="text" control={<Radio size="small" />} label="Enter manually" />
      </RadioGroup>

      {fontInputMode === 'select' ? (
        <Select
          labelId="font-family-label"
          id="fontFamily"
          value={fontFamily}
          onChange={handleFontSelect}
          fullWidth
          variant="outlined"
          size="small"
          sx={{ mb: 1 }}
          data-testid="font-family-select"
        >
          {WEB_SAFE_FONTS.map((font) => (
            <MenuItem key={font} value={font} style={{ fontFamily: font }}>
              {font}
            </MenuItem>
          ))}
        </Select>
      ) : (
        <TextField
          autoFocus
          margin="dense"
          id="fontFamily"
          aria-labelledby="font-family-label"
          type="text"
          fullWidth
          variant="outlined"
          size="small"
          value={fontFamily}
          sx={{ mt: 0, mb: 1 }}
          onChange={handleFontChange}
          data-testid="font-family-input"
        />
      )}

      <Typography variant="subtitle2" gutterBottom sx={{ mt: 2 }} id="monospace-font-family-label">Monospace Font Family</Typography>
      <RadioGroup
        row
        aria-labelledby="monospace-font-family-label"
        name="monospace-font-input-mode-group"
        value={fontInputModeMonospace}
        onChange={handleFontInputModeMonospaceChange}
        sx={{ mb: 1 }}
      >
        <FormControlLabel value="select" control={<Radio size="small" />} label="Select from list" />
        <FormControlLabel value="text" control={<Radio size="small" />} label="Enter manually" />
      </RadioGroup>

      {fontInputModeMonospace === 'select' ? (
        <Select
          labelId="monospace-font-family-label"
          id="fontFamilyMonospace"
          value={fontFamilyMonospace}
          onChange={handleMonospaceFontSelect}
          fullWidth
          variant="outlined"
          size="small"
          sx={{ mb: 1 }}
          data-testid="monospace-font-family-select"
        >
          {WEB_SAFE_MONOSPACE_FONTS.map((font) => (
            <MenuItem key={font} value={font} style={{ fontFamily: font }}>
              {font}
            </MenuItem>
          ))}
        </Select>
      ) : (
        <TextField
          margin="dense"
          id="fontFamilyMonospace"
          aria-labelledby="monospace-font-family-label"
          type="text"
          fullWidth
          variant="outlined"
          size="small"
          value={fontFamilyMonospace}
          sx={{ mt: 0, mb: 1 }}
          onChange={handleMonospaceFontChange}
          data-testid="monospace-font-family-input"
        />
      )}
    </Box>
  );
};

export default FontSettingsTab;
