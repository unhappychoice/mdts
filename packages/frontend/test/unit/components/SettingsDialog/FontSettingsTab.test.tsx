import { render } from '@testing-library/react';
import React from 'react';
import FontSettingsTab from '../../../../src/components/SettingsDialog/FontSettingsTab';

describe('FontSettingsTab', () => {
  it('should render correctly', () => {
    const { asFragment } = render(<FontSettingsTab />);
    expect(asFragment()).toMatchSnapshot();
  });
});
