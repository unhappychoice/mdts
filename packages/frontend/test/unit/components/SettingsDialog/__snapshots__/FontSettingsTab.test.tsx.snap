// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`FontSettingsTab should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-19midj6"
  >
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-1685pvh-MuiTypography-root"
    >
      Font Size
    </h6>
    <span
      class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeSmall css-x7n0q8-MuiSlider-root"
    >
      <span
        class="MuiSlider-rail css-1lmi6fw-MuiSlider-rail"
      />
      <span
        class="MuiSlider-track css-1iqz70e-MuiSlider-track"
        style="left: 0%; width: 0%;"
      />
      <span
        class="MuiSlider-mark MuiSlider-markActive css-81wool-MuiSlider-mark"
        data-index="0"
        style="left: 0%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="1"
        style="left: 7.142857142857143%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="2"
        style="left: 14.285714285714286%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="3"
        style="left: 21.428571428571427%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="4"
        style="left: 28.571428571428573%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="5"
        style="left: 35.714285714285715%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="6"
        style="left: 42.857142857142854%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="7"
        style="left: 50%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="8"
        style="left: 57.142857142857146%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="9"
        style="left: 64.28571428571429%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="10"
        style="left: 71.42857142857143%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="11"
        style="left: 78.57142857142857%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="12"
        style="left: 85.71428571428571%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="13"
        style="left: 92.85714285714286%;"
      />
      <span
        class="MuiSlider-mark css-lq5hqx-MuiSlider-mark"
        data-index="14"
        style="left: 100%;"
      />
      <span
        class="MuiSlider-thumb MuiSlider-thumbSizeSmall MuiSlider-thumbColorPrimary MuiSlider-thumb MuiSlider-thumbSizeSmall MuiSlider-thumbColorPrimary css-171ym8b-MuiSlider-thumb"
        data-index="0"
        style="left: 0%;"
      >
        <input
          aria-label="Font Size"
          aria-labelledby="font-size-slider"
          aria-orientation="horizontal"
          aria-valuemax="24"
          aria-valuemin="10"
          aria-valuenow="10"
          data-index="0"
          max="24"
          min="10"
          step="1"
          style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
          type="range"
          value="10"
        />
        <span
          aria-hidden="true"
          class="MuiSlider-valueLabel css-p0ndrw-MuiSlider-valueLabel"
        >
          <span
            class="MuiSlider-valueLabelCircle"
          >
            <span
              class="MuiSlider-valueLabelLabel"
            >
              10
            </span>
          </span>
        </span>
      </span>
    </span>
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-s0b3wc-MuiTypography-root"
      id="font-family-label"
    >
      Font Family
    </h6>
    <div
      aria-labelledby="font-family-label"
      class="MuiFormGroup-root MuiFormGroup-row MuiRadioGroup-root MuiRadioGroup-row css-1ew1ryv-MuiFormGroup-root"
      role="radiogroup"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall css-1q72ply-MuiButtonBase-root-MuiSwitchBase-root-MuiRadio-root"
        >
          <input
            class="PrivateSwitchBase-input css-gjsmck-MuiSwitchBase-root"
            name="font-input-mode-group"
            type="radio"
            value="select"
          />
          <span
            class="css-9oxshb-MuiRadioButtonIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-10fh2vq-MuiSvgIcon-root-MuiRadioButtonIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-4reyio-MuiSvgIcon-root-MuiRadioButtonIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-1px98du-MuiTypography-root"
        >
          Select from list
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall css-1q72ply-MuiButtonBase-root-MuiSwitchBase-root-MuiRadio-root"
        >
          <input
            class="PrivateSwitchBase-input css-gjsmck-MuiSwitchBase-root"
            name="font-input-mode-group"
            type="radio"
            value="text"
          />
          <span
            class="css-9oxshb-MuiRadioButtonIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-10fh2vq-MuiSvgIcon-root-MuiRadioButtonIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-4reyio-MuiSvgIcon-root-MuiRadioButtonIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-1px98du-MuiTypography-root"
        >
          Enter manually
        </span>
      </label>
    </div>
    <div
      aria-labelledby="font-family-label"
      class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-2lrxdj-MuiFormControl-root-MuiTextField-root"
      data-testid="font-family-input"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl MuiInputBase-sizeSmall css-1iqcefa-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-111x7or-MuiInputBase-input-MuiOutlinedInput-input"
          id="fontFamily"
          type="text"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1nf2c5d-MuiNotchedOutlined-root"
          >
            <span
              aria-hidden="true"
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <h6
      class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-gutterBottom css-s0b3wc-MuiTypography-root"
      id="monospace-font-family-label"
    >
      Monospace Font Family
    </h6>
    <div
      aria-labelledby="monospace-font-family-label"
      class="MuiFormGroup-root MuiFormGroup-row MuiRadioGroup-root MuiRadioGroup-row css-1ew1ryv-MuiFormGroup-root"
      role="radiogroup"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall css-1q72ply-MuiButtonBase-root-MuiSwitchBase-root-MuiRadio-root"
        >
          <input
            class="PrivateSwitchBase-input css-gjsmck-MuiSwitchBase-root"
            name="monospace-font-input-mode-group"
            type="radio"
            value="select"
          />
          <span
            class="css-9oxshb-MuiRadioButtonIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-10fh2vq-MuiSvgIcon-root-MuiRadioButtonIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-4reyio-MuiSvgIcon-root-MuiRadioButtonIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-1px98du-MuiTypography-root"
        >
          Select from list
        </span>
      </label>
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-19opgx6-MuiFormControlLabel-root"
      >
        <span
          class="MuiButtonBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall PrivateSwitchBase-root MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall MuiRadio-root MuiRadio-colorPrimary MuiRadio-sizeSmall css-1q72ply-MuiButtonBase-root-MuiSwitchBase-root-MuiRadio-root"
        >
          <input
            class="PrivateSwitchBase-input css-gjsmck-MuiSwitchBase-root"
            name="monospace-font-input-mode-group"
            type="radio"
            value="text"
          />
          <span
            class="css-9oxshb-MuiRadioButtonIcon-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-10fh2vq-MuiSvgIcon-root-MuiRadioButtonIcon-root"
              data-testid="RadioButtonUncheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
              />
            </svg>
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-4reyio-MuiSvgIcon-root-MuiRadioButtonIcon-root"
              data-testid="RadioButtonCheckedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
              />
            </svg>
          </span>
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-1px98du-MuiTypography-root"
        >
          Enter manually
        </span>
      </label>
    </div>
    <div
      aria-labelledby="monospace-font-family-label"
      class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-2lrxdj-MuiFormControl-root-MuiTextField-root"
      data-testid="monospace-font-family-input"
    >
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-1iqcefa-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-111x7or-MuiInputBase-input-MuiOutlinedInput-input"
          id="fontFamilyMonospace"
          type="text"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-1nf2c5d-MuiNotchedOutlined-root"
          >
            <span
              aria-hidden="true"
              class="notranslate"
            >
              â€‹
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
